### ISSUE TITLE
need a bit more help for making my autocompletion work ...

### ISSUE THREAD

**Issue by jmwatte** [fish-shell/fish-shell#2540]
Still trying to make an autocompletion program, hit a brickwall.
I got
```
    function __fish_beet_needs_command
    set cmd (commandline -opc)
    if [ (count $cmd) -eq 1 -a $cmd[1] = 'beet' ]
        return 0
    end
    return 1
end

function __fish_beet_using_command
    set cmd (commandline -opc)
    if [ (count $cmd) -eq 2 ]
        if [ $argv[1] = $cmd[2] ]
            return 0
        end
    end
    return 1
end

function __fish_beet_use_extras
    set cmd (commandline -opc)
    if [ (count $cmd) -gt 2 -a contains $argv[2] $cmd ] 
        return 0
    end
    return 1
end

set FIELDS acoustid_fingerprint acoustid_id added album album_id albumartist albumartist_credit albumartist_sort albumdisambig 

set GENRES 'Acapella' 'Acid House' 'Acid Jazz' 'Afrobeat' 'Alternative Country' 'Alternative Metal' 'Alternative Rock' 'Ambient' 'Americana' 

complete -f -c beet -n '__fish_beet_needs_command' -a dj -d "send music to a player as a playlist" 
complete -f -c beet -n '__fish_beet_using_command dj' -a "$FIELDS" -d "fields"
complete -f -c beet -n '__fish_beet_use_extras dj genre:' -a "$GENRES" -d "genres"
complete -r -f -c beet -n '__fish_beet_using_command dj' -s A -l args -d "add additional arguments to the command"
complete -r -f -c beet -n '__fish_beet_using_command dj' -s t -l take -d "take this number,or size MB or minutes M"
```
What gives me problems is this call
```
complete -f -c beet -n '__fish_beet_use_extras dj genre:' -a "$GENRES" -d "genres"
```
What I hoped for was that when I type `genre:` somewhere after `beet dj` I would get the autosuggested list from `GENRES`
but I get 
```
test: Expected a combining operator like '-a' at index 5
```
. I've been trying all kind of things in `function __fish_beet_use_extras` but all the wrong things  it seems.
Any help is appreciated and suggestions for beter coding or a better solution too,
thx
jm.

**Comment by faho** [comment]
So, the commandline is supposed to look like this:

```fish
beet dj genre: Alternative\ Rock
```

?

In that case, this line:

```fish
    if [ (count $cmd) -gt 2 -a contains $argv[2] $cmd ] 
```

should be

```fish
    if begin [ (count $cmd) -gt 2]; and contains -- $argv[2] $cmd; end
```

You need the begin/end because our precedence is currently weird. I'd think about if it can always only be argv[2], though - can't there be options before, or other filters?

Also, you don't want to check `$cmd[1] = 'beet'` - fish only executes your completion when beet is somehow invoked anyway, and this breaks wrappers (think aliases).

**Comment by jmwatte** [comment]
Now I got this 
```
function __fish_beet_needs_command
    set cmd (commandline -opc)
    if test (count $cmd) -eq 1
        return 0
    end
    return 1
end

function __fish_beet_using_command
    set cmd (commandline -opc)
    if test (count $cmd) -gt 1
        if test $argv[1] = $cmd[2]
            return 0
        end
    end
    return 1
end

function __fish_beet_use_extra
    set cmd (commandline -opc)
    set needle (count $cmd)
    if test $argv[2]  = $cmd[$needle]
        return 0
    end
    return 1
end
```

and with

```
complete -c beet -n '__fish_beet_needs_command' -a dj -f -d 'send music to a player as a playlist'
complete -c beet -n '__fish_beet_using_command dj' -a "$FIELDS" -f -d "fieldname"
complete -c beet -n '__fish_beet_use_extra dj genre:' -f -A -a '$GENRES' -d 'a genre value'
````

I got most of my problems solved. Except 1.

```
[I] home:Desktop(1)(I)> beet dj genre:                                                                                                                                                jmw@jean
_notags                  (a genre value)  Celtic                   (a genre value)  filesize:              (fieldname)  mb_artistid:             (fieldname)  rg_album_gain:       (fieldname)
Acapella                 (a genre value)  Celtic\ Fusion           (a genre value)  Flamenco           (a genre value)  mb_releasegroupid:       (fieldname)  rg_album_peak:       (fieldname)
Acid\ House              (a genre value)  Chamber\ Pop             (a genre value)  Folk\ Rock         (a genre value)  mb_trackid:              (fieldname)  rg_track_gain:       (fieldname)
Acid\ Jazz               (a genre value)  channels:                    (fieldname)  format:                (fieldname)  media:                   (fieldname)  rg_track_peak:       (fieldname)
…and 42 more rows

```

Inside `genre:` I also get the suggestions from `$FIELDS`. (fieldname). They are hooked up to the `dj` command and even if it is not the end of the world I would rather not have them here. I thought maybe the `-A` option would get me there but that means (I think) that only everything that gets suggested is OK. Not autosuggest just what is in this case `$GENRES`. I get that that is what is supposed to happen because '$FIELDS' are arguments for `dj` so I get them in `genre:` too. But how to prevent that.
Any help, pointers appreciated.
Oh, and if this is not the right place to ask these things, please do say so.
grtz
jm

**Comment by faho** [comment]
>I get that that is what is supposed to happen because '$FIELDS' are arguments for dj so I get them in genre: too. But how to prevent that.

Well, you need to adjust the conditions so the $FIELDS one isn't satisfied when "genre:" is used. This might get a bit tricky.

Are you sure though that it's `genre: Folk`, and not `genre:Folk`? The former strikes me as rather unusual. If you needed to complete the latter, you need to offer `genre:Folk` and other genres, not just `Folk` - cartesian products can be helpful for that (e.g. `genre:$GENRES` if $GENRES is a list variable that contains genres).

> Oh, and if this is not the right place to ask these things, please do say so.

No, it's fine. We have a "question" label for a reason.

**Comment by pickfire** [comment]
@jmwatte Actually you can use the syntax highlighting by using:

    ```fish

**Comment by jmwatte** [comment]
@faho.
Thx. First the tricky one. The plan is to see if the last token on the commanfline is from the $FIELDS and then check if the previous is also from $FIELDS. In that case no go. 
As for the second tip. That one is above my paygrade ☺.I'll have to study up on that one or get a translatipn in laymans term.
@pickfire 
Thx. Picked up a new trick. That one will come in handy.

**Comment by jmwatte** [comment]
OK.
for the tricky one..

```fish
function __fish_beet_using_command
    set cmd (commandline -opc)
    set needle (count $cmd)
    if test $needle -gt 1
        if begin test $argv[1] = $cmd[2];
            and not contains -- $cmd[$needle] $FIELDS; end
                return 0
        end
    end
    return 1
end
```

seems to be doing the job. If I am correct it checks if there is  a $FIELDS member at the last position of
the commandline. If not, you can put one $FIELD member on there.

@faho 
concerning

>Are you sure though that it's genre: Folk, and not genre:Folk? The former strikes me as rather unusual. If you needed to complete the latter, you need to offer genre:Folk and other genres, not just Folk - cartesian products can be helpful for that (e.g. genre:$GENRES if $GENRES is a list variable that contains genres)

I'm not sure I understand you. Well, I don't, actually ;)
But I'm open to understanding.
This program I use [beet] (https://github.com/sampsyo/beets)
-for which I write this autocompletion - does not make a syntactical difference between ``genre: Folk`` or ``genre:Folk`` .It parses it either way and delivers the songs that have a Folk-genre-tag. So I just do ``genre:``  ``TAB`` in fish and I get this list of all the genres.
I'm not a computerprogrammer... just tinkering to get it to work but if you would explain or contextualize what you said I might learn something new.
Anyway.. don't know if I should close this issue or keep it open in case something comes up again.
but faho ... I leave that in your very helpfull hands. Already thx alot .....
.

**Comment by faho** [comment]
>If I am correct it checks if there is a $FIELDS member at the last position of
the commandline

Yes, but you can skip the "needle" stuff - `$list[-1]` is the last element of a list, `$list[-2]` the second-to-last and so on.

>This program I use beet
-for which I write this autocompletion - does not make a syntactical difference between genre: Folk or genre:Folk

Ah, okay. The reason I'm asking is because `genre: Folk` would send it as _two_ arguments, while `genre:Folk` is one argument, so the program needs to handle these two cases a bit differently. Quite often programs don't.

Now, fish's completion is geared towards completing _single, full_ arguments. This means if you offer "Folk" as a possibility, fish won't take it if it is currently completing "genre:" (since it expects something starting with "genre:").

So that means if a program is expecting something of the form "A:B", the completion needs to offer not just "B", but "A:B", for every possible combination of As and Bs. The easiest way to generate that is to use the "cartesian product" (a mathematical term meaning "combine everything with everything"), which looks like `genre:$FIELDS` (if FIELDS is a list).

Try

```fish
set FIELDS Folk {Black,Death,Power,True,Classic,Sludge,Doom,Folk}" Metal"
echo $FIELDS
echo $FIELDS[3]
echo genre:$FIELDS
```

to explore that behavior a bit and perhaps see what I mean.

**Comment by jmwatte** [comment]
Waw, thanks.
I just tried to do that in python but fish has a much beter readable syntax. Yeah, and now I know what you mean.
But the requirements for this are just a list of genres that one has in one's library. Great. Thanks for the insight and the time. Buy you a beer if I run into you. I'll close for now.

**Comment by pickfire** [comment]
@jmwatte I don't think you need that `;`:
```fish
        if begin test $argv[1] = $cmd[2];
```
